examples/petsc/src/petscutils.c

```diff
@@ -167,3 +170,209 @@ PetscErrorCode CreateRestrictionFromPlex(Ceed ceed, DM dm, CeedInt height,
+                          * num_comp];
+
+        interp[q * P + p_ceed] = basis_tabulation->T[0][((face * Q + q) * P * num_comp +
+                                 p_petsc) * num_comp + c];
+        for (CeedInt d = 0; d < dim; d++) {
+          grad[(d * Q + q) * P + p_ceed] = basis_tabulation->T[1][(((
+                                             face * Q + q) * P * num_comp + p_petsc) * num_comp + c) * dim + d];
+        }
+      }
+    }
+
+    // -- Cleanup
+    if (permutation) PetscCall(ISRestoreIndices(permutation, &permutation_indices));
+    PetscCall(ISDestroy(&permutation));
+  }
+
+  // Finally, create libCEED basis
+  CeedBasisCreateH1(ceed, elem_topo, num_comp, P, Q, interp, grad, q_points,
+                    q_weights, basis);
+  PetscCall(PetscFree(q_points));
+  PetscCall(PetscFree(interp));
+  PetscCall(PetscFree(grad));
+
+  PetscFunctionReturn(0);
+}
+
+// -----------------------------------------------------------------------------
+// Get CEED Basis from DMPlex
+// -----------------------------------------------------------------------------
+PetscErrorCode CreateBasisFromPlex(Ceed ceed, DM dm, DMLabel domain_label,
+                                   CeedInt label_value, CeedInt height,
+                                   CeedInt dm_field, CeedBasis *basis) {
+  PetscDS         ds;
+  PetscFE         fe;
+  PetscQuadrature quadrature;
+  PetscBool       is_simplex = PETSC_TRUE;
+  PetscInt        ds_field   = -1;
+
+  PetscFunctionBeginUser;
+
+  // Get element information
+  PetscCall(DMGetRegionDS(dm, domain_label, NULL, &ds));
+  PetscCall(DMFieldToDSField(dm, domain_label, dm_field, &ds_field));
+  PetscCall(PetscDSGetDiscretization(ds, ds_field, (PetscObject *)&fe));
+  PetscCall(PetscFEGetHeightSubspace(fe, height, &fe));
+  PetscCall(PetscFEGetQuadrature(fe, &quadrature));
+
+  // Check if simplex or tensor-product mesh
+  PetscCall(DMPlexIsSimplex(dm, &is_simplex));
+
+  // Build libCEED basis
+  if (is_simplex) {
+    PetscTabulation basis_tabulation;
+    PetscInt        num_derivatives = 1, face = 0;
+
+    PetscCall(PetscFEGetCellTabulation(fe, num_derivatives, &basis_tabulation));
+    PetscCall(BasisCreateFromTabulation(ceed, dm, domain_label, label_value, height,
+                                        face, fe, basis_tabulation, quadrature, basis));
+  } else {
+    PetscDualSpace dual_space;
+    PetscInt       num_dual_basis_vectors;
+    PetscInt       dim, num_comp, P, Q;
+
+    PetscCall(PetscFEGetSpatialDimension(fe, &dim));
+    PetscCall(PetscFEGetNumComponents(fe, &num_comp));
+    PetscCall(PetscFEGetDualSpace(fe, &dual_space));
+    PetscCall(PetscDualSpaceGetDimension(dual_space, &num_dual_basis_vectors));
+    P = num_dual_basis_vectors / num_comp;
+    PetscCall(PetscQuadratureGetData(quadrature, NULL, NULL, &Q, NULL, NULL));
+
+    CeedInt P_1d = (CeedInt)round(pow(P, 1.0 / dim));
+    CeedInt Q_1d = (CeedInt)round(pow(Q, 1.0 / dim));
+
```

Comment:
Here returns Q_1d = 3 which makes sense for `p_degree=3`.
But in main we have `P= p_degree +1`, and `Q=P + q_extra` which gives you 5.

Reply:
See diff above - small bug in older logic